### PR TITLE
exposure analysis with pod selectors

### PR DESCRIPTION
**Pull request by shireenf-ibm**
issue #236 

task:

> - [ ] expanding the exposure analysis to rules with `PodSelector`

### CODE REVIEW

File: pkg/netpol/eval/resources.go [hunk 1]
```diff
@@ -582,29 +583,31 @@ func (pe *PolicyEngine) ConvertPeerNamedPort(namedPort string, peer Peer) (int32
...
 		Namespace: ns,
 		FakePod:   true,
+		Labels:    podLabels,
 	}
 	if err := pe.resolveSingleMissingNamespace(ns, nsLabels); err != nil {
 		return nil, err
 	}
-	if pe.exposureAnalysisFlag && newPod.Name == k8s.RepresentativePodName { // if exposure-analysis and this is not a fake ingress-controller
+	// if exposure-analysis and this is not a fake ingress-controller
+	if pe.exposureAnalysisFlag && strings.HasPrefix(newPod.Name, k8s.RepresentativePodName) {
```
Comment (adisos) [inline pkg/netpol/eval/resources.go:603]: The expression `strings.HasPrefix(newPod.Name, k8s.RepresentativePodName) ` is repeated twice.. better have an explicit function `IsPodRepresentative()`

File: pkg/netpol/eval/resources.go [hunk 2]
```diff
@@ -582,29 +583,31 @@ func (pe *PolicyEngine) ConvertPeerNamedPort(namedPort string, peer Peer) (int32
...
 		// first compute a unique string from labels to be used as a map key
-		keyStrFromNsLabels := k8s.VariantFromLabelsMap(objLabels.NsLabels)
-		// todo : when supporting also pod labels; add also key str from pod labels to the map key
-		if _, ok := pe.representativePeersMap[keyStrFromNsLabels]; ok { // we already have a representative peer with same labels
+		keyStrFromLabels := k8s.VariantFromLabelsMap(objLabels.NsLabels) + k8s.VariantFromLabelsMap(objLabels.PodLabels)
```
Comment (adisos) [inline pkg/netpol/eval/resources.go:605]: add a delimiter for concatenation, to be on the safe side that no two objects are mapped to the same key

File: pkg/netpol/connlist/conns_formatter.go [hunk 3]
```diff
@@ -68,20 +68,41 @@ func formExposureItemAsSingleConnFiled(peerStr string, exposureItem XgressExposu
 	if exposureItem.IsExposedToEntireCluster() {
 		return formSingleExposureConn(peerStr, entireCluster, exposureItem.PotentialConnectivity(), isIngress)
 	}
-	if len(exposureItem.NamespaceLabels()) > 0 {
-		return formSingleExposureConn(peerStr, peerStrWithNsLabels(exposureItem.NamespaceLabels()),
-			exposureItem.PotentialConnectivity(), isIngress)
+	repPeerStr := writeNsLabels(exposureItem.NamespaceLabels())
```
Comment (adisos) [inline pkg/netpol/connlist/conns_formatter.go:71]: rename func `writeNsLabels`, as the name is misleading (it does not write..),
could be `getRepresentativeNamespaceString()` for example

File: pkg/netpol/connlist/conns_formatter.go [hunk 4]
```diff
@@ -68,20 +68,41 @@ func formExposureItemAsSingleConnFiled(peerStr string, exposureItem XgressExposu
...
+	if repPeerStr != "" {
+		repPeerStr += "/"
 	}
-	// @todo handle podLabels
-	return singleConnFields{}
+	repPeerStr += writePodLabels(exposureItem.PodLabels())
```
Comment (adisos) [inline pkg/netpol/connlist/conns_formatter.go:75]: same here for `writePodLabels`

File: pkg/netpol/connlist/conns_formatter.go [hunk 5]
```diff
@@ -68,20 +68,41 @@ func formExposureItemAsSingleConnFiled(peerStr string, exposureItem XgressExposu
 	if exposureItem.IsExposedToEntireCluster() {
 		return formSingleExposureConn(peerStr, entireCluster, exposureItem.PotentialConnectivity(), isIngress)
 	}
-	if len(exposureItem.NamespaceLabels()) > 0 {
-		return formSingleExposureConn(peerStr, peerStrWithNsLabels(exposureItem.NamespaceLabels()),
-			exposureItem.PotentialConnectivity(), isIngress)
+	repPeerStr := writeNsLabels(exposureItem.NamespaceLabels())
+	if repPeerStr != "" {
+		repPeerStr += "/"
```
Comment (adisos) [inline pkg/netpol/connlist/conns_formatter.go:73]: if `/` is not added , does it mean "any namespace" ?

**Review by adisos — CHANGES_REQUESTED** [review]
Looks good!
added few initial comments, did not review the tests yet

**Comment by adisos** [comment]
Missing an example test with exposure to "all namespaces, certain pod labels" ?

File: pkg/netpol/connlist/conns_formatter.go (hunk 4 above)
Comment (shireenf-ibm) [reply pkg/netpol/connlist/conns_formatter.go]: done

File: pkg/netpol/connlist/conns_formatter.go (hunk 3 above)
Comment (shireenf-ibm) [reply pkg/netpol/connlist/conns_formatter.go]: done

File: pkg/netpol/eval/resources.go (hunk 2 above)
Comment (shireenf-ibm) [reply pkg/netpol/eval/resources.go]: done

File: pkg/netpol/eval/resources.go (hunk 1 above)
Comment (shireenf-ibm) [reply pkg/netpol/eval/resources.go]: done

File: pkg/netpol/connlist/conns_formatter.go [hunk 6]
```diff
@@ -86,21 +83,21 @@ const (
...
 )
 
-// writeNsLabels returns a string representation of a potential peer with namespace labels
-func writeNsLabels(nsLabels map[string]string) string {
+// getRepresentativeNamespaceString returns a string representation of a potential peer with namespace labels
+func getRepresentativeNamespaceString(nsLabels map[string]string) string {
 	nsName, ok := nsLabels[common.K8sNsNameLabelKey]
 	if len(nsLabels) == 1 && ok {
 		return nsName
 	}
```
Comment (adisos) [inline pkg/netpol/connlist/conns_formatter.go]: is there a test that covers this in the output?

Comment (shireenf-ibm) [reply pkg/netpol/connlist/conns_formatter.go]: yes, "tests/test_conn_to_all_pods_in_a_new_ns"

**Review by adisos — APPROVED** [review]

**Pull request merged**